File: src/modules/profile/person-profile.controller.ts
import {
  Controller,
  Post,
  Get,
  Param,
  Patch,
  Delete,
  Body,
  UseGuards,
  Req,
  UnauthorizedException,
  Query,
} from '@nestjs/common';
import {
  ApiTags,
  ApiBearerAuth,
  ApiOperation,
  ApiResponse,
  ApiSecurity,
  ApiParam,
} from '@nestjs/swagger';
import { AuthGuard } from '@nestjs/passport';
import { RolesGuard } from 'src/common/guards/roles.guard';
import { ActiveRoleGuard } from 'src/common/guards/active-role.guard';
import { Roles } from 'src/common/decorators/roles.decorator';
import { UserRole } from 'src/enums/user-role.enum';

import {
  CreatePersonProfileDto,
  UpdatePersonProfileDto,
  PersonProfileFilterDto,
} from './dto/person-profile.dto';
import { PersonProfileService } from './person-profile.service';
import { PersonProfile } from './schemas/person-profile.schema';

@ApiTags('person-profiles')
@ApiBearerAuth()
@UseGuards(AuthGuard('jwt'))
@Controller('person-profiles')
export class PersonProfileController {
  constructor(private readonly personProfileService: PersonProfileService) {}

  @Post()
  @ApiSecurity('jwt')
  @ApiOperation({ summary: 'Create a person profile' })
  @ApiResponse({
    status: 201,
    description: 'Person profile created successfully',
  })
  async create(
    @Req() req,
    @Body() createPersonProfileDto: CreatePersonProfileDto,
  ) {
    if (!req.user?.userId) {
      throw new UnauthorizedException('Invalid token');
    }

    createPersonProfileDto.userId = req.user.userId;
    return this.personProfileService.createProfile(createPersonProfileDto);
  }

  @Get()
  @ApiSecurity('jwt')
  @ApiOperation({
    summary: 'Retrieve all person profiles with filtering and pagination',
  })
  @ApiResponse({ status: 200, description: 'List of all person profiles' })
  @UseGuards(AuthGuard('jwt'), ActiveRoleGuard, RolesGuard)
  @Roles(UserRole.ADMIN)
  async findAll(
    @Query() filterDto: PersonProfileFilterDto,
  ): Promise<{ data: PersonProfile[]; total: number }> {
    return this.personProfileService.findAll(filterDto);
  }

  @Get(':id')
  @ApiSecurity('jwt')
  @ApiOperation({ summary: 'Get a person profile by ID' })
  @ApiParam({ name: 'id', required: true })
  @ApiResponse({ status: 200, description: 'Person profile found' })
  async getById(@Param('id') id: string, @Req() req) {
    if (!req.user?.userId) {
      throw new UnauthorizedException('Invalid token');
    }

    return this.personProfileService.findProfileById(id);
  }

  @Get('user/:userId')
  @ApiSecurity('jwt')
  @ApiOperation({ summary: 'Get a person profile by userId' })
  @ApiParam({ name: 'userId', required: true })
  @ApiResponse({ status: 200, description: 'Person profile found' })
  async findByUserId(@Param('userId') userId: string, @Req() req) {
    if (!req.user?.userId) {
      throw new UnauthorizedException('Invalid token');
    }

    return this.personProfileService.findProfileByUserId(userId);
  }

  @Patch(':id')
  @ApiSecurity('jwt')
  @ApiOperation({ summary: 'Update a person profile' })
  @ApiParam({ name: 'id', required: true })
  @ApiResponse({ status: 200, description: 'Profile updated successfully' })
  async update(
    @Param('id') id: string,
    @Body() updateDto: UpdatePersonProfileDto,
    @Req() req,
  ) {
    if (!req.user?.userId) {
      throw new UnauthorizedException('Invalid token');
    }

    return this.personProfileService.updateProfile(
      id,
      updateDto,
      req.user.userId,
    );
  }

  @Delete(':id')
  @ApiSecurity('jwt')
  @ApiOperation({ summary: 'Delete a person profile' })
  @ApiParam({ name: 'id', required: true })
  @ApiResponse({ status: 200, description: 'Profile deleted successfully' })
  async delete(@Param('id') id: string, @Req() req) {
    if (!req.user?.userId) {
      throw new UnauthorizedException('Invalid token');
    }

    return this.personProfileService.deleteProfile(id, req.user.userId);
  }
}

File: RapidaQuickstart.DotNet/Controllers/PersonProfileController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using RapidaQuickstart.DotNet.DTOs;
using RapidaQuickstart.DotNet.Services;
using System.Security.Claims;

namespace RapidaQuickstart.DotNet.Controllers
{
    /// <summary>
    /// Person profile management endpoints
    /// </summary>
    [ApiController]
    [Route("api/person-profiles")]
    public class PersonProfileController : ControllerBase
    {
        private readonly IPersonProfileService _personProfileService;

        public PersonProfileController(IPersonProfileService personProfileService)
        {
            _personProfileService = personProfileService;
        }

        /// <summary>
        /// Create a new person profile
        /// </summary>
        /// <param name="dto">Person profile data</param>
        /// <returns>Created person profile</returns>
        /// <response code="201">Person profile created successfully</response>
        /// <response code="400">Invalid request data</response>
        /// <response code="401">Unauthorized</response>
        [HttpPost]
        [Authorize(Roles = "Person")]
        public async Task<IActionResult> CreateProfile([FromBody] CreatePersonProfileDto dto)
        {
            try
            {
                var userId = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { message = "Unauthorized" });
                }

                dto.UserId = userId;
                var profile = await _personProfileService.CreateProfileAsync(dto);
                return CreatedAtAction(nameof(GetProfileById), new { id = profile.Id }, profile);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get all person profiles with filtering and pagination
        /// </summary>
        /// <param name="filterDto">Filter and pagination parameters</param>
        /// <returns>Paginated list of person profiles</returns>
        /// <response code="200">Returns paginated person profiles</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetAllProfiles([FromQuery] PersonProfileFilterDto filterDto)
        {
            try
            {
                var (data, total) = await _personProfileService.FindAllAsync(filterDto);
                return Ok(new
                {
                    data,
                    total,
                    page = filterDto.Page,
                    limit = filterDto.Limit,
                    totalPages = (int)Math.Ceiling((double)total / filterDto.Limit)
                });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get person profile by ID
        /// </summary>
        /// <param name="id">Profile ID</param>
        /// <returns>Person profile</returns>
        /// <response code="200">Returns person profile</response>
        /// <response code="404">Profile not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet("{id}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetProfileById(string id)
        {
            try
            {
                var profile = await _personProfileService.FindProfileByIdAsync(id);
                if (profile == null)
                {
                    return NotFound(new { message = "Profile not found" });
                }
                return Ok(profile);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get person profile by user ID
        /// </summary>
        /// <param name="userId">User ID</param>
        /// <returns>Person profile</returns>
        /// <response code="200">Returns person profile</response>
        /// <response code="404">Profile not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet("user/{userId}")]
        [Authorize(Roles = "Admin")]
        public async Task<IActionResult> GetProfileByUserId(string userId)
        {
            try
            {
                var profile = await _personProfileService.FindProfileByUserIdAsync(userId);
                if (profile == null)
                {
                    return NotFound(new { message = "Profile not found" });
                }
                return Ok(profile);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Get current user's person profile
        /// </summary>
        /// <returns>Current user's person profile</returns>
        /// <response code="200">Returns person profile</response>
        /// <response code="404">Profile not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpGet("me")]
        [Authorize(Roles = "Person")]
        public async Task<IActionResult> GetMyProfile()
        {
            try
            {
                var userId = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { message = "Unauthorized" });
                }

                var profile = await _personProfileService.FindProfileByUserIdAsync(userId);
                if (profile == null)
                {
                    return NotFound(new { message = "Profile not found" });
                }
                return Ok(profile);
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Update person profile
        /// </summary>
        /// <param name="id">Profile ID</param>
        /// <param name="dto">Updated profile data</param>
        /// <returns>Updated person profile</returns>
        /// <response code="200">Profile updated successfully</response>
        /// <response code="404">Profile not found</response>
        /// <response code="401">Unauthorized</response>
        /// <response code="400">Invalid request data</response>
        [HttpPut("{id}")]
        [Authorize(Roles = "Person")]
        public async Task<IActionResult> UpdateProfile(string id, [FromBody] UpdatePersonProfileDto dto)
        {
            try
            {
                var userId = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { message = "Unauthorized" });
                }

                var profile = await _personProfileService.UpdateProfileAsync(id, dto, userId);
                return Ok(profile);
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (InvalidOperationException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }

        /// <summary>
        /// Delete person profile
        /// </summary>
        /// <param name="id">Profile ID</param>
        /// <returns>Success message</returns>
        /// <response code="200">Profile deleted successfully</response>
        /// <response code="404">Profile not found</response>
        /// <response code="401">Unauthorized</response>
        [HttpDelete("{id}")]
        [Authorize(Roles = "Person")]
        public async Task<IActionResult> DeleteProfile(string id)
        {
            try
            {
                var userId = User.FindFirst("userId")?.Value;
                if (string.IsNullOrEmpty(userId))
                {
                    return Unauthorized(new { message = "Unauthorized" });
                }

                await _personProfileService.DeleteProfileAsync(id, userId);
                return Ok(new { message = "Profile deleted successfully" });
            }
            catch (UnauthorizedAccessException ex)
            {
                return Unauthorized(new { message = ex.Message });
            }
            catch (InvalidOperationException ex)
            {
                return NotFound(new { message = ex.Message });
            }
            catch (Exception ex)
            {
                return BadRequest(new { message = ex.Message });
            }
        }
    }
}
